File: Software/C++/battle.cpp
/*
Compiling and Running in UNIX:

 - Open a terminal
 - Navigate to the same directory as battle.cpp
      cd ~/Programs/jkwalter.github.io/Software/C++
 - Type the following to complie battle.cpp
      g++ battle.cpp
 - Type the following to run battle.cpp
      ./a.out
*/

#include <iostream>
#include <stdlib.h>
#include <time.h>

using namespace std;

class character
{
 private:
   int health;
   int attack;
   int heal;
 public:
   character(void);
   void setAttack(int d1, int d2);
   int getAttack(void);
   void damageValue(int damage);
   int getHealth(void);
   void healValue(int d1, int d2);
   int getHeal(void);
};

class opponent
{
 private:
   int health;
   int attack;
   int heal;
 public:
   opponent(void);
   void setAttack(int d1, int d2);
   int getAttack(void);
   void damageValue(int damage);
   int getHealth(void);
   void healValue(int d1, int d2);
   int getHeal(void);
};

character::character(void)
{
 health = 50;
}
void character::setAttack(int d1, int d2)
{
 if (d1 + d2 == 2 || d1 + d2 == 12)
  {
   attack = 20;
  }
 if (d1 + d2 == 3 || d1 + d2 == 11)
  {
   attack = 14;
  }
 if (d1 + d2 == 4 || d1 + d2 == 10)
  {
   attack = 10;
  }
 if (d1 + d2 == 5 || d1 + d2 == 9)
  {
   attack = 6;
  }
 if (d1 + d2 == 6 || d1 + d2 == 8)
  {
   attack = 6;
  }
 if (d1 + d2 == 7)
  {
   attack = 10;
  }
}
int character::getAttack(void)
{
 return attack;
}
void character::damageValue(int damage)
{
 if (health - damage < 0)
  {
   health = 0;
  }
 else
  {
   health = health - damage;
  }
}
int character::getHealth(void)
{
 return health;
}
void character::healValue(int d1, int d2)
{
 if (d1 + d2 == 2 || d1 + d2 == 12)
  {
   heal = 20;
  }
 if (d1 + d2 == 3 || d1 + d2 == 11)
  {
   heal = 14;
  }
 if (d1 + d2 == 4 || d1 + d2 == 10)
  {
   heal = 10;
  }
 if (d1 + d2 == 5 || d1 + d2 == 9)
  {
   heal = 6;
  }
 if (d1 + d2 == 6 || d1 + d2 == 8)
  {
   heal = 6;
  }
 if (d1 + d2 == 7)
  {
   heal = 10;
  }

 if (health + heal > 50)
  {
   health = 50;
  }
 else
  {
   health = health + heal;
  }
}
int character::getHeal(void)
{
 return heal;
}

opponent::opponent(void)
{
 health = 50;
}
void opponent::setAttack(int d1, int d2)
{
 if (d1 + d2 == 2 || d1 + d2 == 12)
  {
   attack = 20;
  }
 if (d1 + d2 == 3 || d1 + d2 == 11)
  {
   attack = 14;
  }
 if (d1 + d2 == 4 || d1 + d2 == 10)
  {
   attack = 10;
  }
 if (d1 + d2 == 5 || d1 + d2 == 9)
  {
   attack = 6;
  }
 if (d1 + d2 == 6 || d1 + d2 == 8)
  {
   attack = 6;
  }
 if (d1 + d2 == 7)
  {
   attack = 10;
  }
}
int opponent::getAttack(void)
{
 return attack;
}
void opponent::damageValue(int damage)
{
 if (health - damage < 0)
  {
   health = 0;
  }
 else
  {
   health = health - damage;
  }
}
int opponent::getHealth(void)
{
 return health;
}
void opponent::healValue(int d1, int d2)
{
 if (d1 + d2 == 2 || d1 + d2 == 12)
  {
   heal = 20;
  }
 if (d1 + d2 == 3 || d1 + d2 == 11)
  {
   heal = 14;
  }
 if (d1 + d2 == 4 || d1 + d2 == 10)
  {
   heal = 10;
  }
 if (d1 + d2 == 5 || d1 + d2 == 9)
  {
   heal = 6;
  }
 if (d1 + d2 == 6 || d1 + d2 == 8)
  {
   heal = 6;
  }
 if (d1 + d2 == 7)
  {
   heal = 10;
  }

 if (health + heal > 50)
  {
   health = 50;
  }
 else
  {
   health = health + heal;
  }
}
int opponent::getHeal(void)
{
 return heal;
}

int main(void)
{
 character hero;
 opponent villain;
 int option;
 char key;
 int d1;
 int d2;

 srand(time(NULL));

 do
  {
   system("clear");
   cout << "          YOUR HEALTH: " << hero.getHealth() << endl;
   cout << "   OPPONTENT'S HEALTH: " << villain.getHealth() << endl;
   do
    {
     cout << endl;
     cout << "1: Attack" << endl;
     cout << "2: Heal" << endl;
     cout << endl;
     cout << "Enter an option here (1-2): ";
     cin >> option;
    } while (option != 1 && option != 2);
   system("clear");
   switch (option)
    {
     case 1:
            d1 = rand() % 6 + 1;
            d2 = rand() % 6 + 1;
            hero.setAttack(d1, d2);
            villain.damageValue(hero.getAttack());
            cout << "Opponent takes " << hero.getAttack() << " damage." << endl;
            break;
     case 2:
            d1 = rand() % 6 + 1;
            d2 = rand() % 6 + 1;
            hero.healValue(d1, d2);
            cout << "You restored "<< hero.getHeal() << " health." << endl;
            break;
    }
   cout << endl;
   cout << "Enter any key to continue: ";
   cin >> key;
   if (villain.getHealth() != 0)
    {
     system("clear");
     if (villain.getHealth() < 25)
      {
       option = rand() % 10 + 1;
       if (option > 4)
        {
         option = 2;
        }
       else
        {
         option = 1;
        }
      }
     else
      {
       option = rand() % 10 + 1;
       if (option > 3)
        {
         option = 1;
        }
       else
        {
         option = 2;
        }
      }
     switch (option)
      {
       case 1:
              d1 = rand() % 6 + 1;
              d2 = rand() % 6 + 1;
              villain.setAttack(d1, d2);
              hero.damageValue(villain.getAttack());
              cout << "You take " << villain.getAttack() << " damage." << endl;
              break;
       case 2:
              d1 = rand() % 6 + 1;
              d2 = rand() % 6 + 1;
              villain.healValue(d1, d2);
              cout << "Opponent restored " << villain.getHeal() << " health." <<endl;
              break;
      }
     cout << endl;
     cout << "Enter any key to continue: ";
     cin >> key;
    }
   option = 0;
  } while (hero.getHealth() != 0 && villain.getHealth() != 0);

 system("clear");

 if (hero.getHealth() != 0)
  {
   cout << "VICTORY!" << endl;
  }
 else
  {
   cout << "DEFEAT!" << endl;
  }

 cout << endl;
 cout << "          YOUR HEALTH: " << hero.getHealth() << endl;
 cout << "   OPPONTENT'S HEALTH: " << villain.getHealth() << endl;

 return 0;
}

File: Software/C#/battle.cs
/*
Compiling and Running in Microsoft Visual Studio:

 - Open Developer Command Prompt for Visual Studio
 - Navigate to the same directory as battle.cs
      cd "C:\Users\Josh Walter\Documents\Visual Studio 2017\Projects"
 - Type the following to compile battle.cs
      csc battle.cs
 - Type the following to run battle.cs
      battle
*/

class Character
{
	private int myHealth = 0;
	private int attackValue = 0;
	private int healValue = 0;

	public int Health
	{
		get
		{
			return myHealth;
		}
		set
		{
			myHealth = value;
		}
	}

	public int Attack
	{
		get
		{
			return attackValue;
		}
		set
		{
			attackValue = value;
		}
	}

	public int Heal
	{
		get
		{
			return healValue;
		}
		set
		{
			healValue = value;
		}
	}

	public static void Main()
	{
		Character Protagonist = new Character();
		Character Antagonist = new Character();
		System.ConsoleKey Choice;
		System.ConsoleKey Key;
		System.ConsoleKey Enter;
		int Coin = 0;
		int Option = 0;
		int FirstTurn = 0;
		int Dice1;
		int Dice2;
		System.Random Roll = new System.Random();
		System.Random AntagonistDecision = new System.Random();
		System.Random Toss = new System.Random();

		System.Console.Clear();

		Protagonist.Health = 50;
		Antagonist.Health = 50;

		Coin = Toss.Next(2);
		FirstTurn = Coin;

		do
		{
			if (Coin == 0)
			{
				if (Protagonist.Health != 0)
				{
					if (FirstTurn == 0)
					{
						System.Console.WriteLine("   PROTAGONIST HEALTH: {0}", Protagonist.Health);
						System.Console.WriteLine("    ANTAGONIST HEALTH: {0}", Antagonist.Health);
						System.Console.WriteLine();
						System.Console.Write("Enter any key to continue: ");
						Key = System.Console.ReadKey().Key;
						Enter = System.Console.ReadKey().Key;
						System.Console.WriteLine();
						System.Console.WriteLine();
					}
					do
					{
						System.Console.WriteLine("PROTAGONIST OPTIONS");
						System.Console.WriteLine();
						System.Console.WriteLine("   1: Attack");
						System.Console.WriteLine("   2: Heal");
						System.Console.WriteLine();
						System.Console.Write("Enter your option here (1 or 2): ");
						Choice = System.Console.ReadKey().Key;
						Enter = System.Console.ReadKey().Key;
						System.Console.WriteLine();
						System.Console.WriteLine();
					} while (Choice != System.ConsoleKey.D1 && Choice != System.ConsoleKey.D2);
					if (Choice == System.ConsoleKey.D1)
					{
						Option = 1;
					}
					if (Choice == System.ConsoleKey.D2)
					{
						Option = 2;
					}
					switch (Option)
					{
						case 1:
							Dice1 = Roll.Next(6) + 1;
							Dice2 = Roll.Next(6) + 1;
                            if (Dice1 + Dice2 == 2 || Dice1 + Dice2 == 12)
                            {
                                Protagonist.Attack = 20;
                            }
                            if (Dice1 + Dice2 == 3 || Dice1 + Dice2 == 11)
                            {
                                Protagonist.Attack = 14;
                            }
                            if (Dice1 + Dice2 == 4 || Dice1 + Dice2 == 10)
                            {
                                Protagonist.Attack = 10;
                            }
                            if (Dice1 + Dice2 == 5 || Dice1 + Dice2 == 9)
                            {
                                Protagonist.Attack = 6;
                            }
                            if (Dice1 + Dice2 == 6 || Dice1 + Dice2 == 8)
                            {
                                Protagonist.Attack = 6;
                            }
                            if (Dice1 + Dice2 == 7)
                            {
                                Protagonist.Attack = 10;
                            }
                            if (Antagonist.Health < Protagonist.Attack)
                            {
                                Antagonist.Health = 0;
                            }
                            else
                            {
                                Antagonist.Health = Antagonist.Health - Protagonist.Attack;
                            }
                            System.Console.WriteLine("Protagonist deals {0} damage.", Protagonist.Attack);
                            break;
                        case 2:
                            Dice1 = Roll.Next(6) + 1;
                            Dice2 = Roll.Next(6) + 1;
                            if (Dice1 + Dice2 == 2 || Dice1 + Dice2 == 12)
                            {
                                Protagonist.Heal = 20;
                            }
                            if (Dice1 + Dice2 == 3 || Dice1 + Dice2 == 11)
                            {
                                Protagonist.Heal = 14;
                            }
                            if (Dice1 + Dice2 == 4 || Dice1 + Dice2 == 10)
                            {
                                Protagonist.Heal = 10;
                            }
                            if (Dice1 + Dice2 == 5 || Dice1 + Dice2 == 9)
                            {
                                Protagonist.Heal = 6;
                            }
                            if (Dice1 + Dice2 == 6 || Dice1 + Dice2 == 8)
                            {
                                Protagonist.Heal = 6;
                            }
                            if (Dice1 + Dice2 == 7)
                            {
                                Protagonist.Heal = 10;
                            }
                            if (Protagonist.Health + Protagonist.Heal > 50)
                            {
                                Protagonist.Health = 50;
                            }
                            else
                            {
                                Protagonist.Health = Protagonist.Health + Protagonist.Heal;
                            }
                            System.Console.WriteLine("Protagonist restores {0} health.", Protagonist.Heal);
                            break;
                    }
                    System.Console.WriteLine();
                    System.Console.Write("Enter any key to continue: ");
                    Key = System.Console.ReadKey().Key;
                    Enter = System.Console.ReadKey().Key;
                    System.Console.WriteLine();
                    System.Console.WriteLine();
                    Coin = 1;
                }
            }
            if (Coin == 1)
            {
                if (Antagonist.Health != 0)
                {
                    if (FirstTurn == 1)
                    {
                        System.Console.WriteLine("   PROTAGONIST HEALTH: {0}", Protagonist.Health);
                        System.Console.WriteLine("    ANTAGONIST HEALTH: {0}", Antagonist.Health);
                        System.Console.WriteLine();
                        System.Console.Write("Enter any key to continue: ");
                        Key = System.Console.ReadKey().Key;
                        Enter = System.Console.ReadKey().Key;
                        System.Console.WriteLine();
                        System.Console.WriteLine();
                    }
                    if (Antagonist.Health < 20)
                    {
                        Option = AntagonistDecision.Next(10);
                        if (Option < 4)
                        {
                            Option = 1;
                        }
                        else
                        {
                            Option = 2;
                        }
                    }
                    else
                    {
                        Option = AntagonistDecision.Next(10);
                        if (Option < 3)
                        {
                            Option = 2;
                        }
                        else
                        {
                            Option = 1;
                        }
                    }
                    switch (Option)
                    {
                        case 1:
                            Dice1 = Roll.Next(6) + 1;
                            Dice2 = Roll.Next(6) + 1;
                            if (Dice1 + Dice2 == 2 || Dice1 + Dice2 == 12)
                            {
                                Antagonist.Attack = 20;
                            }
                            if (Dice1 + Dice2 == 3 || Dice1 + Dice2 == 11)
                            {
                                Antagonist.Attack = 14;
                            }
                            if (Dice1 + Dice2 == 4 || Dice1 + Dice2 == 10)
                            {
                                Antagonist.Attack = 10;
                            }
                            if (Dice1 + Dice2 == 5 || Dice1 + Dice2 == 9)
                            {
                                Antagonist.Attack = 6;
                            }
                            if (Dice1 + Dice2 == 6 || Dice1 + Dice2 == 8)
                            {
                                Antagonist.Attack = 6;
                            }
                            if (Dice1 + Dice2 == 7)
                            {
                                Antagonist.Attack = 10;
                            }
                            if (Protagonist.Health < Antagonist.Attack)
                            {
                                Protagonist.Health = 0;
                            }
                            else
                            {
                                Protagonist.Health = Protagonist.Health - Antagonist.Attack;
                            }
                            System.Console.WriteLine("Antagonist deals {0} damage.", Antagonist.Attack);
                            break;
                        case 2:
                            Dice1 = Roll.Next(6) + 1;
                            Dice2 = Roll.Next(6) + 1;
                            if (Dice1 + Dice2 == 2 || Dice1 + Dice2 == 12)
                            {
                                Antagonist.Heal = 20;
                            }
                            if (Dice1 + Dice2 == 3 || Dice1 + Dice2 == 11)
                            {
                                Antagonist.Heal = 14;
                            }
                            if (Dice1 + Dice2 == 4 || Dice1 + Dice2 == 10)
                            {
                                Antagonist.Heal = 10;
                            }
                            if (Dice1 + Dice2 == 5 || Dice1 + Dice2 == 9)
                            {
                                Antagonist.Heal = 6;
                            }
                            if (Dice1 + Dice2 == 6 || Dice1 + Dice2 == 8)
                            {
                                Antagonist.Heal = 6;
                            }
                            if (Dice1 + Dice2 == 7)
                            {
                                Antagonist.Heal = 10;
                            }
                            if (Antagonist.Health + Antagonist.Heal > 50)
                            {
                                Antagonist.Health = 50;
                            }
                            else
                            {
                                Antagonist.Health = Antagonist.Health + Antagonist.Heal;
                            }
                            System.Console.WriteLine("Antagonist restores {0} health.", Antagonist.Heal);
                            break;
                    }
                    System.Console.WriteLine();
                    System.Console.Write("Enter any key to continue: ");
                    Key = System.Console.ReadKey().Key;
                    Enter = System.Console.ReadKey().Key;
                    System.Console.WriteLine();
                    System.Console.WriteLine();
                    Coin = 0;
				}
			}
		} while (Protagonist.Health != 0 && Antagonist.Health != 0);

		if (Antagonist.Health == 0)
		{
			System.Console.WriteLine("VICTORY!");
		}
		if (Protagonist.Health == 0)
		{
			System.Console.WriteLine("DEFEAT!");
		}

		System.Console.WriteLine();
		System.Console.WriteLine("   PROTAGONIST HEALTH: {0}", Protagonist.Health);
		System.Console.WriteLine("    ANTAGONIST HEALTH: {0}", Antagonist.Health);
	}
}
